src/useForm.tsx

```diff
@@ -90,13 +94,37 @@ function createFields<T>(
           // This avoids issues related to stale state when multiple setValue functions are
           // invoked within the same render cycle -- e.g.
           // const onClick = () => { date.day.setValue(...); date.day.setValue(...) }:
-          setValue: (updatedValue: any) => {
-            setState(currentState =>
-              produce(currentState, updatedState => {
-                set(updatedState, [...path, 'value'], updatedValue)
-                set(updatedState, [...path, 'touched'], true)
-              })
-            )
+          setValue: <
+            U extends
+              | SetValueOptions
+              | EmptySetValueOptions = EmptySetValueOptions
+          >(
+            updatedValue: any,
+            options: SetValueOptions | undefined = undefined
+          ): SetValueReturnType<U> => {
+            const isValid = new Promise<boolean>(resolve => {
+              setState(currentState =>
+                produce(currentState, updatedState => {
```

Comment:
Nice! I hadn't heard of `immer` before, but this is handy for reducing boilerplate.

Reply:
Yea, the "use an intelligent proxy" pattern is useful in other contexts too. For example, I used this in Beyonce's [update API](https://github.com/ginger-io/beyonce#update) -- which allows you to update a Dynamo item using the same "pretend I'm mutating the real object" syntax: 

```TypeScript
const updatedAuthor = await beyonce.update(AuthorModel.key({ id: "1" }), (author) => {
  // author is a proxy object here, so we can "pretend" to mutate it, even though we haven't actually read it from the DB
 // which is nifty for "partial update" requests coming in from a client. 
  author.name = "Jack London",
  author.details.description = "American novelist"
  delete author.details.someDeprecatedField
})
```